pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go

```diff
@@ -468,19 +488,194 @@ func (m panicNodeMutator) Generate(
 		// are compatible with a dead node, so we immediately restart the node after the panic.
 		if validEndStep == nil {
+			restartStep = cutStep
 			addRestartStep = cutStep.InsertBefore(restartNodeStep{planner.currentContext.System.Descriptor.Nodes[0], targetNode, planner.rt, restartDesc})
 		} else {
-			addRestartStep = validEndStep.
-				RandomStep(rng).
+			restartStep = validEndStep.RandomStep(rng)
+			addRestartStep = restartStep.
 				Insert(rng, restartNodeStep{planner.currentContext.System.Descriptor.Nodes[0], targetNode, planner.rt, restartDesc})
 		}
 
+		failureContextSteps, _ := validStartStep.CutBefore(func(s *singleStep) bool {
+			return s == restartStep[0]
+		})
+		failureContextSteps.MarkNodesUnavailable(true, false)
+
 		mutations = append(mutations, addPanicStep...)
 		mutations = append(mutations, addRestartStep...)
 	}
 
-	return mutations
+	return mutations, nil
+}
+
+func (m panicNodeMutator) DisabledDeploymentModes() map[DeploymentMode]struct{} {
+	return nil
+}
+
+func (m panicNodeMutator) SupportsLocalRun() bool {
+	return true
+}
+
+type networkPartitionMutator struct{}
+
+func (m networkPartitionMutator) Name() string { return failures.IPTablesNetworkPartitionName }
+
+func (m networkPartitionMutator) Probability() float64 {
+	return 0.3
+}
+
+func (m networkPartitionMutator) Generate(
+	rng *rand.Rand, plan *TestPlan, planner *testPlanner,
+) ([]mutation, error) {
+	var mutations []mutation
+	upgrades := randomUpgrades(rng, plan)
+	idx := newStepIndex(plan)
+	nodeList := planner.currentContext.System.Descriptor.Nodes
+
+	failure := failures.GetFailureRegistry()
+	f, err := failure.GetFailer(planner.cluster.Name(), failures.IPTablesNetworkPartitionName, planner.logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get failer for %s: %w", failures.IPTablesNetworkPartitionName, err)
+	}
+
+	for _, upgrade := range upgrades {
+		possiblePointsInTime := upgrade.
+			Filter(func(s *singleStep) bool {
+				// We don't want to set up a partition concurrently with other steps, and inserting
+				// before a concurrent step causes the step to run concurrently with that step, so
+				// we filter out any concurrent steps. We don't want to set up a partition while
+				// nodes are unavailable, as that could cause the cluster to lose quorum,
+				//	so we filter out steps with unavailable nodes.
+				var unavailableNodes bool
+				if planner.isMultitenant() {
+					unavailableNodes = s.context.Tenant.hasUnavailableNodes || s.context.System.hasUnavailableNodes
+				} else {
+					unavailableNodes = s.context.System.hasUnavailableNodes
+				}
+				return s.context.System.Stage >= InitUpgradeStage && !idx.IsConcurrent(s) && !unavailableNodes
+			})
+
+		stepToPartition := possiblePointsInTime.RandomStep(rng)
+		hasInvalidConcurrentStep := false
+		var firstStepInConcurrentBlock *singleStep
+
+		isInvalidRecoverStep := func(s *singleStep) bool {
+			// Restarting a node in the middle of a network partition has a chance of
```

Comment:
Maybe it would be easier to visualize a scenario if you gave an example of a partition + a restarted node to show how it loses quorum?